pkg/jit/jitdump.go

```diff
@@ -0,0 +1,460 @@
+	}
+	if err := p.read(&jr.CodeSize); err != nil {
+		return nil, fmt.Errorf("codeSize: %w", err)
+	}
+	if err := p.read(&jr.CodeIndex); err != nil {
+		return nil, fmt.Errorf("codeIndex: %w", err)
+	}
+
+	return jr, nil
+}
+
+func (p *jitDumpParser) parseJRCodeDebugInfo(prefix *JRPrefix) (*JRCodeDebugInfo, error) {
+	jr := &JRCodeDebugInfo{Prefix: prefix}
+
+	if err := p.read(&jr.CodeAddr); err != nil {
+		return nil, fmt.Errorf("codeAddr: %w", err)
+	}
+	if err := p.read(&jr.NREntry); err != nil {
+		return nil, fmt.Errorf("nrEntry: %w", err)
+	}
+
+	size := jrCodeDebugInfoFixedSize + debugEntryFixedSize*int(jr.NREntry)
+
+	jr.Entries = make([]*DebugEntry, jr.NREntry)
+	for i := uint64(0); i < jr.NREntry; i++ {
+		jr.Entries[i] = &DebugEntry{}
+
+		if err := p.read(&jr.Entries[i].Addr); err != nil {
+			return nil, fmt.Errorf("entries[%d].addr: %w", i, err)
+		}
+		if err := p.read(&jr.Entries[i].Lineno); err != nil {
+			return nil, fmt.Errorf("entries[%d].lineno: %w", i, err)
+		}
+		if err := p.read(&jr.Entries[i].Discrim); err != nil {
+			return nil, fmt.Errorf("entries[%d].discrim: %w", i, err)
+		}
+
+		var err error
+		jr.Entries[i].Name, err = p.readString()
+		if err != nil {
+			return nil, fmt.Errorf("entries[%d].name: %w", i, err)
+		}
+		size += len(jr.Entries[i].Name) + 1 // +1 accounts for trimmed null termination
+	}
+
+	// Discard padding if any
+	if _, err := p.buf.Discard(int(jr.Prefix.TotalSize) - size); err != nil {
+		return nil, fmt.Errorf("failed to discard JIT Code Debug Info record padding: %w", err)
+	}
+
+	return jr, nil
+}
+
+func (p *jitDumpParser) parseJRCodeUnwindingInfo(prefix *JRPrefix) (*JRCodeUnwindingInfo, error) {
+	jr := &JRCodeUnwindingInfo{Prefix: prefix}
+
+	jr.Prefix = prefix
+
+	if err := p.read(&jr.UnwindingSize); err != nil {
+		return nil, fmt.Errorf("unwindingSize: %w", err)
+	}
+	if err := p.read(&jr.EHFrameHDRSize); err != nil {
+		return nil, fmt.Errorf("ehFrameHDRSize: %w", err)
+	}
+	if err := p.read(&jr.MappedSize); err != nil {
+		return nil, fmt.Errorf("mappedSize: %w", err)
+	}
+
+	jr.UnwindingData = make([]byte, jr.UnwindingSize)
+	if _, err := io.ReadAtLeast(p.buf, jr.UnwindingData, len(jr.UnwindingData)); err != nil {
+		return nil, fmt.Errorf("unwindingData: %w", err)
+	}
+
+	// Discard padding if any
+	if _, err := p.buf.Discard(int(jr.Prefix.TotalSize) - jrCodeUnwindingInfoFixedSize - int(jr.UnwindingSize)); err != nil {
+		return nil, fmt.Errorf("failed to discard JIT Code Unwinding Info record padding: %w", err)
+	}
+
+	return jr, nil
+}
+
+func (p *jitDumpParser) parse() (*JITDump, error) {
+	dump := &JITDump{}
+	var err error
+	dump.Header, err = p.parseJITHeader()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read JIT dump header: %w", isUnexpectedIOError(err))
+	}
+
+	dump.CodeLoads = make([]*JRCodeLoad, 0)
+	dump.CodeMoves = make([]*JRCodeMove, 0)
+	dump.DebugInfo = make([]*JRCodeDebugInfo, 0)
+	dump.UnwindingInfo = make([]*JRCodeUnwindingInfo, 0)
```

Comment:
Should we accept these as arguments to move the allocations to the caller? cc @kakkoyun

Reply:
Yes, we can get the whole `*JITDump` as an argument and leave the allocation to the caller. As a second iteration, we can consider using a pool if it creates a problem.

Reply:
Good suggestion!